### ISSUE TITLE
Any suggestions for debugging?

### ISSUE THREAD

**Issue by xuchen** [gnieh/tiscaf#2]
My server runs fine on my Linux machine. When I ship the code to another one running Linux, it runs fine too. When I ship the code to one running Windows 7, both firefox/IE keep trying to access http://localhost:8080 or 127.0.0.1:8080 but they never load the page (script access enabled in IE, Windows firewall isn't blocking Java either).

I've followed the tutorial and used / for path separators for Windows. Is there a way to find out why it didn't work on Windows when I'm writing the code on Linux? If not, and I have to setup a development environment on Windows, then where in the code should I look into? Thanks!

**Comment by satabin** [comment]
Hi,

I am not sure that this problem is tiscaf specific, but rather related to scala. Try to track down where your program hangs with your usual debugging methods for scala to see where the problem lies (first of all, does your request ever reaches the server?). If it turns out that this is a bug in tiscaf, then you can file a ticket with the specific problem described and potentially a fix. I am not aware of any such problem in tiscaf, but have no windows machine to test. Didn't you hard code the path in the server (something like `/home/user/tiscaf`)?
You'll need to give more information to see if this problem is specific to tiscaf

**Comment by xuchen** [comment]
Thanks! I managed to get a Windows machine and spotted the problem by trying TimeServer in the example folder, which behaved the same as did my own program. The problem is in ResourceLet.scala, in the return statement of function resolvePath(tk: HTalk):

```
new java.io.File(path).getCanonicalPath
```

In specifically the TimeServer example, path == "//", then there was an exception:

java.io.IOException: Invalid argument
    at java.io.WinNTFileSystem.canonicalize0(Native Method)
    at java.io.Win32FileSystem.canonicalize(Unknown Source)
    at java.io.File.getCanonicalPath(Unknown Source)
    at tiscaf.let.ResourceLet$class.resolvePath(ResourceLet.scala:...)
......

I'm not very familiar with Windows paths, but I suppose this should be easy to fix?

BTW whatever exceptions are thrown, they are caught in scala.concurrent.impl.Future and not printed, which is very annoying 'cause you've no idea what has gone wrong. Anyway to override this?

Xuchen

**Comment by satabin** [comment]
For the first part of your message, this is probably standard Java on how to deal with platform dependent separator in paths.

For the last part, can you please file a ticket for it?

**Comment by xuchen** [comment]
For the first part, could you please fix it? I mean, the TimeServer example fails under Windows. I suppose there will be other complaints about it when a Windows user tries it.

Sure! Filing the ticket now.

**Comment by satabin** [comment]
This should be fixed, see commit @2638f8418d0f76efea92f26d9e4853d9c2804ce8 (sorry for the wrong ticket number in commit message...)

**Comment by xuchen** [comment]
Thanks for trying but it doesn't seem to be fixed yet. I tried it with the TimeServer example and got the "404" error. No exception is thrown but path is not right.

Specifically, suppose I'm running  the TimeServer example in some folder on D, in the following code in ResourceLet.scala (function act()):

```
  val path = resolvePath(tk)

  val fullPathAndStream = ("" +: indexes)
    .map(idx => (path + { if (idx.length == 0) "" else "/" + idx }).replace("//", "/"))
    .map { fullPath => (fullPath, getResource(fullPath)) }
    .find(_._2 != null)
```

path here is "D:\" and fullPath here is "D:\/index.html". "D:\" doesn't feel right, since it's not on my classpath. The "\/" in fullPath should be an easy fix though.

Try again? ;-)

**Comment by satabin** [comment]
Well, as I said before, I have no windows machine, if you know what the problem is, and how to resolve it on your windows machine, please submit a pull request. Howerver, the `getResource` method accepts a path, which is not of form `D:\\...` (see tha java doc http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getResource%28java.lang.String%29)

Where does the `D:\` comes from in your code? I don't see in the tiscaf code why you have this, this comes probably from your code. You maybe misunderstood the meaning of `dirRoot` which is not a filesystem path, but a path in classpath (as defined in the javadoc)

**Comment by xuchen** [comment]
OK Finally I got both a Windows and Mac to compare side-by-side. In the resolvePath(tk: HTalk) function of ResourceLet.scala, the return value is:

```
new java.io.File(path).getCanonicalPath
```

I found out that I got path="/" in, and after calling getCanonicalPath, I got "D:\" out, then the rest is screwed up by this "D:\".

The easiest fix that worked both for TimeServer and my own code is replacing

```
new java.io.File(path).getCanonicalPath
```

with just:

```
path
```

Then problem solved. Even if it's not a perfect one, it's working on all my Windows/Linux/Mac machines.
